opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/instrument.py

```diff
@@ -176,6 +180,47 @@ def record(
             Measurement(amount, self, attributes)
         )
 
+    def time(self):
+        return _Timer(self)
+
+
+class _Timer:
```

Comment:
Smells like an antipattern:

We should not add a `time` method to the SDK `Histogram` class, that class should represent an implementation of the API `Histogram` class which does not include a `time` method. What we are trying to do here is to add a new functionality, a new type of histogram that can time an operation.

We should not allow the user to repeatedly associate attributes to a time measurement. A time measurement is just a specialized version of `record` that measures an operation time (where an operation is a bunch of code inside a `with` context or a decorated function). We should only allow attributes to be associated once with the time measurement when we call `time` in the same way that we only allow attributes to be associated with another non-time measurement when we call `record`.

The current implementation uses [`with self._new_timer()`](
https://github.com/open-telemetry/opentelemetry-python/pull/2827/files#diff-17a679cd897e8a5f1f65cc6034945a0714aaf212644445e44ffe04c192d0de32R219) which does not allow the user to associate attributes to a time measurement using a decorated function which should be allowed to make it functionally equivalent with using the timing-capable histogram as a context manager with `with`.

I suggest this instead:

``` python
from functools import wraps
from time import sleep
from unittest import TestCase
from timeit import default_timer


class Histogram:

    def __init__(self):
        self._records = []

    def record(self, amount, attributes):
        self._records.append((amount, attributes))


class TimingHistogram(Histogram):

    def __init__(self):
        super().__init__()
        self._start = None
        self._attributes = None

    def __enter__(self):
        self._start = default_timer()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.record(
            max(round((default_timer() - self._start) * 1000), 0),
            self._attributes
        )
        self._attributes = None

    def time(self, attributes=None):
        self._attributes = attributes
        return self

    def __call__(self, function):
        @wraps(function)
        def wrapped(*args, **kwargs):
            with self.time(attributes=self._attributes):
                return function(*args, **kwargs)

        return wrapped


class TestSuite(TestCase):

    def setUp(self):
        self.timing_histogram = TimingHistogram()

    def test_context_manager(self):

        with self.timing_histogram.time(attributes={"a": "b"}):
            sleep(1)

        self.assertEqual(self.timing_histogram._records, [(1000, {"a": "b"})])

    def test_decorator(self):

        @self.timing_histogram.time(attributes={"c": "d"})
        def function():
            sleep(2)

        function()
        self.assertEqual(self.timing_histogram._records, [(2000, {"c": "d"})])
```

Reply:
> The timing operation is just a convenience method for the users. I do not think there should be a separate type that just adds this capability. Among other reasons, there is only one way to obtain the histogram instrument using the meter API `meter.create_histogram(...)`. A new type means an addition of new API to obtain timing histogram types which isn't really adding any value.

The fact that there is only an API function `create_histogram` is a good point. The issue I see here is that we are adding an "API" in the SDK by adding the `time` function. So, an app that uses the `time` function is now tied to that specific SDK which is not good of course. The _ideal_ thing would be that this `time` functionality is defined in the spec as part of the API, then we implement it in the SDK, etc, etc. like usual. That of course is a long and time consuming thing to do.

I wonder now about a more general question, should we define API-level functionality in the SDK? :thinking:

Reply:
> We can allow optional list of attributes be passed for @time.

Yes, agreed, that is what I suggest in my implementation above :+1:

Reply:
Ok, in summary, these are my concerns:

1. We are adding an "API" in the SDK when we add `time` to the public interface of the SDK `Histogram` class, and that can make the application code dependent on a specific SDK. Since this is an "extra" feature that we (otel Python developers) are adding outside of any spec requirement, shouldn't it be in a util library instead?
2. We try to allow the setting of attributes using both a decorator and a `set_attributes` function. They are not exactly equivalent since the former only allows for attributes to be set once and the latter multiple times. If you are ok with this its fine, I want to make sure we are aware of this difference.

Reply:
>shouldn't it be in a util library instead?

How does this look like? I am trying to understand how we add this capability without adding anything to `Histogram` class.

>We try to allow the setting of attributes using both a decorator and a set_attributes function. They are not exactly equivalent since the former only allows for attributes to be set once and the latter multiple times. If you are ok with this its fine, I want to make sure we are aware of this difference.

Part of it's behaviour is just mimicking the span. They might not have all attributes upfront when using context manager when this `set_attributes` is helpful.

Reply:
> > shouldn't it be in a util library instead?
> 
> How does this look like? I am trying to understand how we add this capability without adding anything to `Histogram` class.

Something like this would work:

``` python
from functools import wraps
from time import sleep
from unittest import TestCase
from timeit import default_timer


class Histogram:

    def __init__(self):
        self._records = []

    def record(self, amount, attributes):
        self._records.append((amount, attributes))


class _TimingHistogram:

    def __init__(self, histogram):
        super().__init__()
        self._start = None
        self._attributes = None
        self._histogram = histogram

    def __enter__(self):
        self._start = default_timer()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self._histogram.record(
            max(round((default_timer() - self._start) * 1000), 0),
            self._attributes
        )
        self._attributes = None

    def time(self, attributes=None):
        self._attributes = attributes
        return self

    def __call__(self, function):
        @wraps(function)
        def wrapped(*args, **kwargs):
            with self.time(attributes=self._attributes):
                return function(*args, **kwargs)

        return wrapped


def timing_histogram(histogram):

    return _TimingHistogram(histogram)


class TestSuite(TestCase):

    def setUp(self):
        self.timing_histogram = timing_histogram(Histogram())

    def test_context_manager(self):

        with self.timing_histogram.time(attributes={"a": "b"}):
            sleep(1)

        self.assertEqual(
            self.timing_histogram._histogram._records, [(1000, {"a": "b"})]
        )

    def test_decorator(self):

        @self.timing_histogram.time(attributes={"c": "d"})
        def function():
            sleep(2)

        function()
        self.assertEqual(
            self.timing_histogram._histogram._records, [(2000, {"c": "d"})]
        )
```

Reply:
Ok. We can do something like that.